main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1694 @@
+      debugln("-".repeat(80))
+      printEquations()
+      debugln()
+    }
+
+    private def phase1UnitPropagation(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 1: Unit Propagation")
+      debugln("    (resolves all equations of the form: x = c where x is a var and c is const)")
+      debugln("-".repeat(80))
+      val s = propagateUnit(currentEqns, currentSubst)
+      updateState(s)
+      printEquations()
+      printSubstitution()
+      debugln()
+    }
+
+    private def phase2VarPropagation(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 2: Variable Propagation")
+      debugln("    (resolves all equations of the form: x = y where x and y are vars)")
+      debugln("-".repeat(80))
+      val s = propagateVars(currentEqns, currentSubst)
+      updateState(s)
+      printEquations()
+      printSubstitution()
+      debugln()
+    }
+
+    private def phase3VarAssignment(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 3: Variable Assignment")
+      debugln("    (resolves all equations of the form: x = t where x is free in t)")
+      debugln("-".repeat(80))
+      val s = varAssignment(currentEqns, currentSubst)
+      updateState(s)
+      printEquations()
+      printSubstitution()
+      debugln()
+    }
+
+    private def phase4SVE(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 4: Boolean Unification")
+      debugln("    (resolves all remaining equations using SVE.)")
+      debugln("-".repeat(80))
+      val newSubst = boolUnifyAllPickSmallest(currentEqns)
+      currentEqns = Nil
+      currentSubst = newSubst @@ currentSubst
+      printSubstitution()
+      debugln()
+    }
+
+    /**
+      * Updates the internal state with the given list of pending equations and partial substitution `l`.
+      *
+      * Simplifies the pending equations. Throws [[ConflictException]] if an unsolvable equation is detected.
+      */
+    private def updateState(l: (List[Equation], BoolSubstitution)): Unit = {
+      val (nextEqns, nextSubst) = l
+      currentEqns = simplify(nextEqns)
+      currentSubst = nextSubst
+    }
+
+    /**
+      * Verifies that the current substitution is a valid solution to the original equations `l`.
+      *
+      * Note: Does not make sense to call before the equation system has been fully solved.
+      *
+      * Throws a [[ConflictException]] if an equation is not solved by the current substitution.
+      */
+    private def verifySolution(): Unit = {
+      verify(currentSubst, l)
+    }
+
+    /**
+      * Checks that current substitution has not grown too large according to the defined threshold.
+      */
+    private def verifySolutionSize(): Unit = {
+      val size = currentSubst.size
+      if (size > Threshold) {
+        throw TooComplexException(size)
+      }
+    }
+
+    private def printEquations(): Unit = {
+      debugln(s"Equations (${currentEqns.size}):")
+      debugln(format(currentEqns))
```

Comment:
Seems fishy when debugln is not by-name - computing the string and then checking

Reply:
Good point!

Reply:
Fixed.